usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -1297,6 +1300,141 @@ CK_RV ep11tok_pkey_add_protkey_attr_to_tmpl(TEMPLATE *tmpl)
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Convert secure key to protected key. */
+    if (ep11_data->mk_change_active) {
+        /* Try to get CKA_IBM_OPAQUE_REENC, ignore if it fails */
+        template_attribute_get_non_empty(key_obj->template,
+                                         CKA_IBM_OPAQUE_REENC, &skey_reenc_attr);
+    }
+
+retry:
+    rc = ep11tok_pkey_skey2pkey(tokdata, session, skey_attr, skey_reenc_attr,
+                                &pkey_attr, xts_mode);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("protkey creation failed, rc=0x%lx\n", rc);
+        goto unlock;
+    }
+
+    /*
+     * In case of XTS, check if the wkvp's of the two keys are identical.
+     * An LGR could have happened between the creation of the two keys.
+     */
+    if (xts_mode) {
+        vp_offset1 = pkey_attr->ulValueLen / 2 - PKEY_MK_VP_LENGTH;
+        vp_offset2 = pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH;
+        if (memcmp((CK_BYTE *)pkey_attr->pValue + vp_offset1,
+                   (CK_BYTE *)pkey_attr->pValue + vp_offset2,
+                   PKEY_MK_VP_LENGTH) != 0) {
+            num_retries++;
+            if (num_retries < PKEY_CONVERT_KEY_RETRIES) {
+                TRACE_DEVEL("%s vp of xts key 1 does not match with vp of "
+                            "xts key 2, retry %d of %d ...\n",
+                            __func__, num_retries, PKEY_CONVERT_KEY_RETRIES);
+                goto retry;
+            }
+            rc = CKR_FUNCTION_FAILED;
+            goto unlock;
+        }
+    }
+
+    /*
+     * Save new protkey attr in key_obj. This happens only in memory,
+     * works also for r/o sessions.
+     */
+    rc = template_update_attribute(key_obj->template, pkey_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed, rc=0x%lx\n", rc);
+        goto unlock;
+    }
+
+    /* If we have an r/w session, also save obj to disk. */
+    if (ep11tok_pkey_session_ok_for_obj(session, key_obj)) {
+        if (object_is_token_object(key_obj)) {
+            rc = object_mgr_save_token_object(tokdata, key_obj);
+            if (rc != CKR_OK) {
+                TRACE_ERROR("Could not save token obj to repository, rc=0x%lx.\n", rc);
+                goto unlock;
+            }
+        }
+    }
+
+    /* Update wkvp in EP11 private data. */
+    if (pthread_mutex_lock(&ep11_data->pkey_mutex)) {
+        TRACE_ERROR("%s Failed to lock pkey lock\n", __func__);
+        rc = CKR_CANT_LOCK;
+        goto unlock;
+    }
+    memcpy(&ep11_data->pkey_mk_vp,
+           (CK_BYTE *)pkey_attr->pValue + pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH,
+           PKEY_MK_VP_LENGTH);
+    if (pthread_mutex_unlock(&ep11_data->pkey_mutex)) {
+        TRACE_ERROR("%s Failed to lock pkey lock\n", __func__);
+    }
+
+    /* Pass back new protkey. Need to do this before unlocking the obj. */
+    memcpy(protkey, pkey_attr->pValue, pkey_attr->ulValueLen);
```

Comment:
same as for the CCA token: Check here for length, instead of at the beginning of the function.

Reply:
fixed.